lib/mongoid/traversable.rb

```diff
@@ -53,19 +55,37 @@ class << self
 
       def discriminator_value=(value)
         value ||= self.name
+        self._clear_types if self.respond_to?(:_clear_types)
         add_discriminator_mapping(value)
         super
       end
     end
 
+    # Module used for prepending the discriminator_value method.
+    #
+    # A separate module was needed because the subclasses of this class
+    # need to be manually prepended with the discriminator_value and can't
+    # rely on being a class_attribute because the .discriminator_value
+    # method is overriden by every subclass in the inherited method.
+    # 
+    # @api private
+    module DiscriminatorRetrieval
+
+      # Get the name on the reading side if the discriminator_value is nil
+      def discriminator_value
+        super || self.name
+      end
+    end
+
     included do
       class_attribute :discriminator_key, instance_accessor: false
       class_attribute :discriminator_value, instance_accessor: false
             
       class << self
         delegate :discriminator_key, to: ::Mongoid
         prepend DiscriminatorAssignment
-        
+        prepend DiscriminatorRetrieval
```

Comment:
I believe this does not need a prepend, so `include` should suffice?

Reply:
I don't understand what this means?

Reply:
How do I use `include` to achieve this functionality?